### ISSUE TITLE
[Bug][Memory] Paged allocator double free can alias live KV pages

### ISSUE THREAD

**Issue by PCBismarck** [LMCache/LMCache#5005]
Suggested labels: `bug`, `ref_count`

## Summary

`PagedTensorMemoryAllocator` accepts the same `TensorMemoryObj` more than once
in `free_blocks`. Subsequent allocations can therefore return the same Python
object and physical page to multiple callers that all believe they own a live,
independent KV allocation.

Writing through either allocation silently overwrites the other allocation's
KV data.

This can be reached through both direct duplicate `free()` calls and duplicate
`ref_count_down()` calls.

## Affected revision

Reproduced against `dev` at `16df1bbb35c958e8353ddeeac66b4e6f89c64800`.

## Root cause

There are two cooperating problems.

First, `TensorMemoryObj.ref_count_down()` changes a negative reference count
back to zero, but then continues into the normal zero-count release condition:

```python
self.meta.ref_count -= 1
if self.meta.ref_count < 0:
    # Logs "Double free occurred somewhere."
    self.meta.ref_count = 0

if self.meta.ref_count == 0 and ...:
    self.parent_allocator.free(self)
```

Calling `ref_count_down()` twice on an object whose count is one therefore
calls the parent allocator twice.

Second, `PagedTensorMemoryAllocator.free()` checks `is_valid()` but does not
invalidate the object. The relevant call is commented out:

```python
if not memory_obj.is_valid():
    return

self.free_blocks.append(memory_obj)
# memory_obj.invalidate()
```

The second `free()` consequently appends the same object to `free_blocks`
again.

With four physical pages, the state can become:

```text
initial free queue:       [P0, P1, P2, P3]
allocate P0:              [P1, P2, P3]
free P0 twice:            [P1, P2, P3, P0, P0]
next five allocations:     P1  P2  P3  P0  P0
```

The allocator now reports five live allocations backed by only four physical
pages.

## Reproduction

This reproduction uses the real allocator and needs no model or GPU:

```python
import torch

from lmcache.v1.memory_management import MemoryFormat
from lmcache.v1.memory_allocators.paged_tensor_memory_allocator import (
    PagedTensorMemoryAllocator,
)

shape = torch.Size([16, 512])
dtype = torch.float16
page_bytes = 16 * 512 * 2
buffer = torch.empty(4 * page_bytes, dtype=torch.uint8)

allocator = PagedTensorMemoryAllocator(
    buffer,
    [shape],
    [dtype],
    MemoryFormat.KV_2LTD,
)

obj = allocator.allocate(shape, dtype)
assert obj is not None
assert obj.get_ref_count() == 1

obj.ref_count_down()
obj.ref_count_down()

live = []
for _ in range(5):
    allocated = allocator.allocate(shape, dtype)
    assert allocated is not None
    live.append(allocated)

left, right = live[-2:]
assert left.raw_data.data_ptr() == right.raw_data.data_ptr()

left.tensor.fill_(3.0)
assert right.tensor.flatten()[0].item() == 3.0
```

Observed output from the full reproducer:

```text
Ref count of MemoryObj 0 is negative: -1. Double free occurred somewhere.
page(b)=0x36015d80 page(c)=0x36015d80 -> ALIASED (same page)
wrote 3.0 through b; reading c gives 3.0
```

Calling `allocator.free(obj)` twice directly produces the same result.

## Actual behavior

- The free queue can grow beyond the allocator's physical page count.
- Multiple live allocations can have the same `data_ptr`.
- KV writes through one live allocation silently modify another allocation.
- Debug allocation counters can also become negative or inconsistent.

## Expected behavior

- A physical page must occur at most once in the free pool.
- Duplicate release must be rejected or be an idempotent no-op before the page
  is enqueued again.
- Every simultaneously live allocation must have unique physical ownership.
- Allocation/free accounting must remain bounded by physical capacity.

## Fix considerations

Uncommenting `memory_obj.invalidate()` alone is not sufficient. The allocation
path resets metadata and `ref_count`, but does not restore `valid = True`, so a
normally recycled page would remain invalid after its first reuse.

A complete fix should define an explicit allocated/free state transition and
apply it consistently to `allocate`, `batched_allocate`, `free`, and
`batched_free`. `ref_count_down()` should also stop before calling the parent
allocator when the count was already zero.

Regression tests should verify:

1. duplicate `free()` does not increase free capacity;
2. duplicate `ref_count_down()` does not enqueue a page twice;
3. allocating physical capacity plus one returns `None`;
4. all simultaneously live allocations have unique `data_ptr` values;
5. a normally freed page becomes valid again when reallocated.

## Related issue

#1365 reports negative reference counts, but does not describe the paged free
queue accepting the same page twice, over-allocation beyond physical capacity,
or aliasing between live KV allocations.

## Environment

- OS: Linux 6.6.87.2 under WSL2, x86_64
- Python: 3.12.14
- PyTorch: 2.13.0+cu130
- Reproduction buffer: CPU tensor; no CUDA execution required
